### PR TITLE
Enhance MapStream

### PR DESCRIPTION
**Pull request by jodastephen**
- Add ability to get inverse
- Add ability to map key and value together

### CODE REVIEW

File: modules/collect/src/test/java/com/opengamma/strata/collect/MapStreamTest.java [hunk 1]
```diff
@@ -141,6 +152,14 @@ public void mapKeysAndValuesToValues() {
     assertThat(result).isEqualTo(expected);
   }
 
+  @Test
+  public void mapBoth() {
+    Map<String, Short> expected =
+        ImmutableMap.of("one - 1", (short) 0, "two - 1", (short) 1, "three - 1", (short) 2, "four - 1", (short) 3);
+    Map<String, Short> result = MapStream.of(map).mapBoth((k, v) -> entry(k + " - 1", (short) (v - 1))).toMap();
```
Comment (wjnicholson) [inline modules/collect/src/test/java/com/opengamma/strata/collect/MapStreamTest.java:159]: Why all the (short) casting?

**Review by wjnicholson — APPROVED** [review]

File: modules/collect/src/test/java/com/opengamma/strata/collect/MapStreamTest.java (hunk 1 above)
Comment (jodastephen) [reply modules/collect/src/test/java/com/opengamma/strata/collect/MapStreamTest.java:159]: To demonstrate that the type of the result can be changed

Comment (wjnicholson) [reply modules/collect/src/test/java/com/opengamma/strata/collect/MapStreamTest.java:159]: What type is being changed? You're going from `Map<String, Short> -> Map<String, Short>`

Comment (jodastephen) [reply modules/collect/src/test/java/com/opengamma/strata/collect/MapStreamTest.java:159]: You can write `short s = 1` to convert the literal `1`. But you cannot convert that way to a capital S `Short` - the `1` is interpreted as an `int`, hence the cast.

Comment (wjnicholson) [reply modules/collect/src/test/java/com/opengamma/strata/collect/MapStreamTest.java:159]: Ahah, got you. Isn't that due to `Guavate.entry` doing the boxing for you?

**Pull request merged**